File: vaild-perfect-square/vaild-perfect-square.py
class Solution:
    # 二分查找
    def isPerfectSquare(self, num: int) -> bool:
        left, right = 1, num
        while 0 < 1:
            mid = (right + left) // 2
            if num < mid**2:
                right = mid
            elif num > mid**2:
                left = mid
            else:
                return True
            if right - left <= 1:
                return False

    # 牛顿近似法
    def isPerfectSquare2(self, num):
        ret = 1
        for i in range(1000):
            ret = (ret + num / ret) // 2
            if ret**2 == num:
                return True
        return False

File: leetcode/vaild-perfect-square/vaild-perfect-square.js
/**
 * @param {number} num
 * @return {boolean}
 */
// 执行用时 : 140 ms
var isPerfectSquare = function (num) {
  var left = 0,
    right = num

  while (left <= right) {
    var mid = left + parseInt((right - left) / 2)
    var pow = mid * mid
    if (pow === num)
      return true
    else if (pow < num)
      left = mid + 1
    else
      right = mid - 1
  }
  return false
};

// 执行用时 : 140 ms
var isPerfectSquare2 = function (num) {
  var a = 0,
    b = num

  while (a <= b) {
    var mid = (a + b) >> 1,
      ans = mid * mid

    if (ans > num)
      b = mid - 1
    else if (ans < num)
      a = mid + 1
    else
      return true
  }
  return false
};